File: go/BinSearch.go
package main

import (
	"fmt"
	"bufio"
	"os"
)

var stdIn = bufio.NewScanner(os.Stdin)

func main(){
	fmt.Println("要素数：")
	var num int = nextInt()
	x := [num]int{}

	fmt.Println("昇順に入力してください。")

	fmt.Println("x[0]: ")
	x[0] = nextInt()

	for i := 1; i < nnum; i++ {
		for x[i] < x[i - 1] {
			println("x[" + i + "]:")
			x[i] = nextInt()
		}
	}

	print("探す値：")
	ky := nextInt()
	idx := binSearch(x, num, ky)

	if idx == -1 {
		println("その値の要素は存在しません。")
	}
	if idx > -1 {
		fmt.Println("その値はx[" + idx + "]にあります。")
	}
}

func nextInt() int {
	    stdIn.Scan()
    i, e := strconv.Atoi(sc.Text())
    if e != nil {
        panic(e)
    }

    return i
}

func binSearch(a []int, n int, key int) int {
	pl := 0
	pr := n - 1

	for pl <= pr {
		var pc int = (pl + pr) / 2

		if a[pc] == key {
			return pc;
		}
		if a[pc] < key {
			pl := pc + 1
		}
		if a[pc] > key {
			pr := pc - 1
		}
	}

	return -1
}

File: kotlin/BinSearch.kt
class BinSearch {
    fun main() {
        print("要素数:")
        val num: Int = readLine()?.toInt() ?: 0
        val x: IntArray(num)

        println("昇順に入力してください。")

        print("x[0]: ")
        x[0] = readLine()?.toInt() ?: 0

        for (i in 0 until num) {
            do {
                print("x[" + i + "]: ")
                x[i] = readLine()?.toInt() ?: 0
            } while (x[i] < x[i - 1])
        }

        print("探す値：")
        val ky: Int = readLine()?.toInt() ?: 0
        val idx: Int = binSearch(x, num, ky)

        if (idx == -1)
            print("その値の要素は存在しません。")
        else
            print("その値はx[" + idx + "]にあります。")
    }

    fun binSearch(a: IntArray, n: Int, key: Int) Int {
        var pl: Int = 0
        var pr: Int = n - 1

        do {
            var pc: Int = (pl + pr) / 2
            if (a[pc] == key) return pc
            if (a[pc] < key)
                pl = pc + 1
            if (a[pc] > key)
                pr = pc + 1
        } while (pl <= pr)

        return -1
    }
}